demo/app/constants.py

```diff
@@ -3,38 +3,34 @@
+    "c4per_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.000008784978490439244,"range":[-1,1],"std_dev":0.0002543762675486505,"sum":129.70260620117188,"valued":"CONTINUOUS"},
+    "cell_area":{"data_type":"FLOAT","has_nan":False,"mean":54771.609375,"range":[8915.4794921875,77276.703125],"std_dev":18437.73046875,"sum":808655454208,"valued":"CONTINUOUS"},
+    "change":{"data_type":"FLOAT","has_nan":False,"mean":0.5,"range":[0,1],"std_dev":0.1,"sum":7382067,"valued":"CONTINUOUS"},
+    "pastr":{"data_type":"FLOAT","has_nan":False,"mean":0.04077955335378647,"range":[0,1],"std_dev":0.10672948509454727,"sum":602074.8125,"valued":"CONTINUOUS"},
+    "pastr_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.00026207335758954287,"range":[-1,1],"std_dev":0.005082341376692057,"sum":3869.28662109375,"valued":"CONTINUOUS"},
+    "primf":{"data_type":"FLOAT","has_nan":False,"mean":0.19610066711902618,"range":[0,1],"std_dev":0.35063520073890686,"sum":2895256.75,"valued":"CONTINUOUS"},
+    "primf_diff":{"data_type":"FLOAT","has_nan":False,"mean":-0.0009334315545856953,"range":[-0.850843608379364,0],"std_dev":0.004068289417773485,"sum":-13781.3095703125,"valued":"CONTINUOUS"},
+    "primn":{"data_type":"FLOAT","has_nan":False,"mean":0.2566087543964386,"range":[0,1],"std_dev":0.3646445870399475,"sum":3788606.5,"valued":"CONTINUOUS"},
+    "primn_diff":{"data_type":"FLOAT","has_nan":False,"mean":-0.001117548905313015,"range":[-0.936556875705719,0],"std_dev":0.005212769843637943,"sum":-16499.642578125,"valued":"CONTINUOUS"},
+    "range":{"data_type":"FLOAT","has_nan":False,"mean":0.15799088776111603,"range":[0,1],"std_dev":0.28534045815467834,"sum":2332598.75,"valued":"CONTINUOUS"},
+    "range_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.00040018701110966504,"range":[-1,1],"std_dev":0.011220048181712627,"sum":5908.4150390625,"valued":"CONTINUOUS"},
+    "secdf":{"data_type":"FLOAT","has_nan":False,"mean":0.10117984563112259,"range":[0,1],"std_dev":0.2359693944454193,"sum":1493832.875,"valued":"CONTINUOUS"},
+    "secdf_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.0006275310879573226,"range":[-1,1],"std_dev":0.004725911188870668,"sum":9264.9541015625,"valued":"CONTINUOUS"},
+    "secdn":{"data_type":"FLOAT","has_nan":False,"mean":0.08007288724184036,"range":[0,1],"std_dev":0.18958471715450287,"sum":1182206.875,"valued":"CONTINUOUS"},
+    "secdn_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.0004467177495826036,"range":[-1,1],"std_dev":0.009596005082130432,"sum":6595.4013671875,"valued":"CONTINUOUS"},
+    "urban":{"data_type":"FLOAT","has_nan":False,"mean":0.0025856471620500088,"range":[0,1],"std_dev":0.021832581609487534,"sum":38174.84375,"valued":"CONTINUOUS"},
+    "urban_diff":{"data_type":"FLOAT","has_nan":False,"mean":0.00004831538171856664,"range":[-0.15093612670898438,0.1676577627658844],"std_dev":0.0006846132455393672,"sum":713.3348388671875,"valued":"CONTINUOUS"}
+}
```

Comment:
Here I realized I forgot to update the fields statistics with updated values from the entire world vs. just West Europe. Previously we were getting slightly skewed prescriptions because the context was being scaled just into the range of West Europe's values. This could also have been why extrapolation was doing poorly - as the UK doesn't have some of the more extreme land uses.

Reply:
Ranges are used only for prescription training, to specify the search space for **actions**.
- Values for land usage (**Context**) are between [0, 1], where 0 means this type of land is not used at all and 1 means this type of land represents 100% of the cell's surface.
- Diffs (**Actions**) should be between [-1, 1], where -1 means 100% of this land type is not used anymore, and 1 means 100% of the land has been converted to that type.

I see a few diffs that are not between [-1, 1], but they are for land types we don't prescribe: `urban`, `primn` and `primf`. So the prescriptors you trained should be fine. Still, you might want to update all the ranges, to be consistent.